src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts

```diff
@@ -52,44 +115,255 @@ export class NotebookKernelSourceSelector implements INotebookKernelSourceSelect
     ) {
-        const quickPickItems = this.kernelFinder.registered.map(this.toQuickPickItem);
+        const items: KernelSourceQuickPickItem[] = [];
+        const allKernelFinders = this.kernelFinder.registered;
+
+        const localKernelFinder = allKernelFinders.find((finder) => finder.id === 'local');
+        if (localKernelFinder) {
+            // local kernel spec and python env finder
+            items.push({
+                type: KernelSourceQuickPickType.LocalKernelSpecAndPythonEnv,
+                label: 'Local Kernels & Python Environments',
+                detail: DataScience.pickLocalKernelTitle(),
+                kernelFinderInfo: localKernelFinder
+            });
+        }
+
+        // Manual remote server kernel finder
+        items.push({
+            type: KernelSourceQuickPickType.LocalServer,
+            label: 'Existing Jupyter Server',
+            detail: DataScience.jupyterSelectURINewDetail()
+        });
+
+        // 3rd party remote server uri providers
+        const providers = await this.uriProviderRegistration.getProviders();
+        providers.forEach((p) => {
+            items.push({
+                type: KernelSourceQuickPickType.ServerUriProvider,
+                label: p.displayName ?? p.id,
+                detail: `Connect to Jupyter servers from ${p.displayName ?? p.id}`,
+                provider: p
+            });
+        });
+
         const selectedSource = await multiStep.showQuickPick<
-            KernelFinderQuickPickItem,
-            IQuickPickParameters<KernelFinderQuickPickItem>
-        >({ items: quickPickItems, placeholder: '', title: DataScience.kernelPickerSelectSourceTitle() });
+            KernelSourceQuickPickItem,
+            IQuickPickParameters<KernelSourceQuickPickItem>
+        >({
+            items: items,
+            placeholder: '',
+            title: DataScience.kernelPickerSelectSourceTitle()
+        });
 
         if (selectedSource) {
-            // Got a source, now get the kernel
-            state.source = selectedSource.kernelFinderInfo;
-            return this.getKernel.bind(this, notebookType);
+            switch (selectedSource.type) {
+                case KernelSourceQuickPickType.LocalKernelSpecAndPythonEnv:
+                    state.source = selectedSource.kernelFinderInfo;
+                    return this.getKernel.bind(this, async () => {
+                        return this.getMatchingControllers(selectedSource.kernelFinderInfo, notebookType);
+                    });
+                case KernelSourceQuickPickType.LocalServer:
+                    return this.getLocalServers.bind(this, notebookType);
+                case KernelSourceQuickPickType.ServerUriProvider:
+                    return this.getRemoteServersFromProvider.bind(this, selectedSource.provider, notebookType);
+                default:
+                    break;
+            }
+        }
+    }
+
+    private async getLocalServers(
```

Comment:
what makes them local? line 200 suggests that they are remote

Reply:
The naming might be misleading but local here means they are manually provided jupyter servers, but not servers created by uri provider.

I think it makes good sense to change the name here.